examples/07_end_to_end/wikimedia.ipynb

```diff
@@ -0,0 +1,547 @@
+    "    def _run(self):\n",
+    "        servernames = dict([])\n",
+    "        while self._running:\n",
+    "            for item in self._source:\n",
+    "                if item.event == 'message':\n",
+    "                    try:\n",
+    "                        change = json.loads(item.data)\n",
+    "                    except ValueError:\n",
+    "                        pass\n",
+    "                    else:\n",
+    "                        # discard canary events\n",
+    "                        # WMF Data Engineering team produces artificial 'canary' events into \n",
+    "                        # each stream multiple times an hour. The presence of these canary\n",
+    "                        # events in a stream allow us to differentiate between a broken event\n",
+    "                        # stream, and an empty one.\n",
+    "                        # We will also filter bot-generated events\n",
+    "                        if change['meta']['domain'] == 'canary' or re.search('bot', change['user'], re.IGNORECASE):\n",
+    "                            continue\n",
+    "                        timestamp = change['meta']['dt']\n",
+    "                        event = f\"{timestamp}:: {change['user']} edited {change['title']}\"\n",
+    "                        servername = change['server_name']\n",
+    "                        # Manually \"tick\" this edge\n",
+    "                        self.push_tick(\n",
+    "                            WikiData(\n",
+    "                                servername=servername,\n",
+    "                                servername_count=0,\n",
+    "                                timestamp=timestamp,\n",
+    "                                event=event\n",
+    "                            )\n",
+    "                        )\n",
+    "\n",
+    "# Create the graph-time representation of our adapter\n",
+    "FetchWikiData = py_push_adapter_def(\"FetchWikiData\", FetchWikiDataAdapter, csp.ts[WikiData], url=str)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "af924b1b-3734-42c8-8978-79f5c88ffb54",
+   "metadata": {},
+   "source": [
+    "Now, we define the graph and run the engine."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9f105395-b9d5-4fa4-b752-48e6d2618fae",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "@csp.node\n",
+    "def update_widget(wiki_event: csp.ts[WikiData], widget: PerspectiveWidget):\n",
+    "    if csp.ticked(wiki_event):\n",
+    "        widget.update({\n",
+    "            \"servername\": [wiki_event.servername],\n",
+    "            \"servername_count\": [wiki_event.servername_count],\n",
+    "            \"timestamp\": [wiki_event.timestamp],\n",
+    "            \"event\": [wiki_event.event],\n",
+    "        })\n",
+    "\n",
+    "@csp.node\n",
+    "def compute_server_count(wiki_event: csp.ts[WikiData], servernames: dict) -> csp.ts[WikiData]:\n",
```

Comment:
Note sure where my comment on this went...
dont pass servernames in as an argument, it should be a state variable that only exists within this node
```
with csp.state():
    s_servernames = {}

...
```

Reply:
Thanks! Just fixed it.